lua/apisix/plugins/wolf-rbac.lua

```diff
@@ -0,0 +1,376 @@
+    local url = ctx.var.uri
+    local action = ctx.var.request_method
+    local clientIP = core.request.get_ip(ctx)
+    local permItem = {action = action, url = url, clientIP = clientIP}
+
+    local rbac_token = fetch_rbac_token(ctx)
+    if rbac_token == nil then
+        core.log.info("no permission to access ", core.json.delay_encode(permItem), ", need login!")
+        return 401, {message = "Missing rbac token in request"}
+    end
+
+    local tokenInfo = parse_rbac_token(rbac_token)
+    core.log.info("token info: ", core.json.delay_encode(tokenInfo))
+    if tokenInfo.err then
+        return 401, {message = 'invalid rbac token: parse failed'}
+    end
+
+
+    local appid = tokenInfo.appid
+    local wolf_token = tokenInfo.wolf_token
+    permItem.appid = appid
+    permItem.wolf_token = wolf_token
+
+    local consumer_conf = consumer.plugin(plugin_name)
+    if not consumer_conf then
+        return 401, {message = "Missing related consumer"}
+    end
+
+    local consumers = core.lrucache.plugin(plugin_name, "consumers_key",
+            consumer_conf.conf_version,
+            create_consume_cache, consumer_conf)
+
+    core.log.info("------ consumers: ", core.json.delay_encode(consumers))
+    local consumer = consumers[appid]
+    if not consumer then
+        core.log.error("consumer [", appid, "] not found")
+        return 401, {message = "Invalid appid in rbac token"}
+    end
+    core.log.info("consumer: ", core.json.delay_encode(consumer))
+    local server = consumer.auth_conf.server
+
+    local url = ctx.var.uri
+    local action = ctx.var.request_method
+    local clientIP = core.request.get_ip(ctx)
+    local permItem = {appid = appid, action = action, url = url, clientIP = clientIP, wolf_token = wolf_token}
+
+    local res = check_url_permission(server, appid, action, url, clientIP, wolf_token)
+    core.log.info(" check_url_permission(", core.json.delay_encode(permItem), ") res: ",core.json.delay_encode(res))
+
+    local username = nil
+    local nickname = nil
+    if type(res.userInfo) == 'table' then
+        local userInfo = res.userInfo
+        core.response.set_header("X-UserId", userInfo.id)
+        core.response.set_header("X-Username", userInfo.username)
+        core.response.set_header("X-Nickname", ngx.escape_uri(userInfo.nickname) or userInfo.username)
+        ctx.userInfo = userInfo
+        username = userInfo.username
+        nickname = userInfo.nickname
+    end
+
+    if res.status == 200 then
+        ---
+    else
+        -- no permission.
+        core.log.error(" check_url_permission(", core.json.delay_encode(permItem), ") failed, res: ",core.json.delay_encode(res))
+        return 401, {message = res.err, username = username, nickname = nickname}
+    end
+    core.log.info("hit wolf-rbac rewrite")
+end
+
+local function get_args()
+    local args, err
+    ngx.req.read_body()
+    if string.find(ngx.req.get_headers()["Content-Type"] or "","application/json", 0) then
```

Comment:
We can turn off the pattern matching feature by using the optional fourth argument plain.

`> = string.find("Hello Lua user", "%su", 1, true) -- turn on plain searches, now not found`

http://lua-users.org/wiki/StringLibraryTutorial

Reply:
`ctx.var.http_content_type` you can make a try in this way which is faster.

Reply:
how to get the `ctx` in a api function?